examples/MQ/pixelSimSplit/src/devices/FairMQChunkMerger.cxx

```diff
@@ -120,21 +119,23 @@ bool FairMQChunkMerger::MergeData(fair::mq::Parts& parts, int /*index*/)
                 continue;   //  want only MCTrack array to renumber tracks and get track shifts...
             //  LOG(info) << "BEFORE ADDING, TCA \"" << tca_elem->GetName() << "\" has " <<
             //  tca_elem->GetEntries() << " entries.";
-            TClonesArray* arrayToAdd;
             for (auto& it = fRet.first; it != fRet.second; ++it) {
                 if (it->second.first == fMCSplitEventHeader->GetChunkStart())
                     continue;
                 if (strcmp(tca_elem->GetName(), it->second.second->GetName()) == 0) {
                     trackShift.push_back(tca_elem->GetEntries());
-                    arrayToAdd = dynamic_cast<TClonesArray*>(it->second.second);
+                    TClonesArray* arrayToAdd = dynamic_cast<TClonesArray*>(it->second.second);
                     for (int iobj = 0; iobj < arrayToAdd->GetEntries(); ++iobj) {
                         FairMCTrack* temp = dynamic_cast<FairMCTrack*>(arrayToAdd->At(iobj));
                         if (temp->GetMotherId() >= 0) {
                             temp->SetMotherId(temp->GetMotherId() + trackShift.back());
                         }
                     }
-                    tca_elem->AbsorbObjects(arrayToAdd);
                     //      LOG(info) << "FOUND ONE!, TCA has now " << tca_elem->GetEntries() << " entries.";
+                    if (!arrayToAdd)
+                        LOG(warning) << "Did not find TCA with name \"" << tca_elem->GetName() << "\"";
+                    else
+                        tca_elem->AbsorbObjects(arrayToAdd);
```

Comment:
Line 127 reads `arrayToAdd` may be a `nullptr` (if the dynamic_cast fails), but line 128 derefences `arrayToAdd` unconditionally which is a bug. See also [C.146](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#c146-use-dynamic_cast-where-class-hierarchy-navigation-is-unavoidable) for more details on the transformation in the following example:

```cpp
if (auto mcTracks = dynamic_cast<TClonesArray*>(it->second.second)) {
    for (auto i = 0; i < mcTracks->GetEntries(); ++i) {
        if (auto mcTrack = dynamic_cast<FairMCTrack*>(mcTracks->At(i)) {
            auto const motherId = mcTrack->GetMotherId();
            if (motherId >= 0) {
                mcTrack->SetMotherId(motherId + trackShift.back());
            }
        }
    }
    tca_elem->AbsorbObjects(mcTracks);
}
```

On a side note: Prefer explicitly braced scope blocks, omitting the braces for one-line blocks is error-prone. (https://clang.llvm.org/extra/clang-tidy/checks/readability-braces-around-statements.html)

Reply:
Just discussed a bit with @ChristianTackeGSI. [cppreference.com](https://en.cppreference.com/w/cpp/language/dynamic_cast) contradicts C.146's recommendation of performing a downcast via `static_cast` in cases where it is guaranteed to work by context. I think this optimization is fine and e.g. the inner cast in above's example could be argued to be such a safe downcast because that is how `TClonesArray` is defined (if I am not mistaken).

However, I want to point out, if you use a `dynamic_cast` and then ignore the potential `nullptr` result is imho a code smell (at least without additional clarifying comments).